EXPLORING EBAY CAR SALES

The data set is of used cars from ebay Kleinanzeigen gotten from dataquest. The columns of the dataset are 20 and are described as follows:


dateCrawled - When this ad was first crawled. All field-values are taken from this date.
name - Name of the car.
seller - Whether the seller is private or a dealer.
offerType - The type of listing
price - The price on the ad to sell the car.
abtest - Whether the listing is included in an A/B test.
vehicleType - The vehicle Type.
yearOfRegistration - The year in which the car was first registered.
gearbox - The transmission type.
powerPS - The power of the car in PS.
model - The car model name.
kilometer - How many kilometers the car has driven.
monthOfRegistration - The month in which the car was first registered.
fuelType - What type of fuel the car uses.
brand - The brand of the car.
notRepairedDamage - If the car has a damage which is not yet repaired.
dateCreated - The date on which the eBay listing was created.
nrOfPictures - The number of pictures in the ad.
postalCode - The postal code for the location of the vehicle.
lastSeenOnline - When the crawler saw this ad last online.


OBJECTIVE

The aim is to clean the data and and do a basic analysis to find out some interesting trends such as most expensive car brands, top selling brands and attempt to find out if theres is some correlation between variables.

LOADING & READING DATA

We start by importing the libraries needed.

In [4]:
pip install plotly

You should consider upgrading via the '/dataquest/system/env/python3/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [64]:
import pandas as pd
import numpy as np

from matplotlib.cbook import delete_masked_points
from matplotlib import pyplot as pyplot

import seaborn as sns
from pandas.plotting import scatter_matrix
from pandas.plotting import autocorrelation_plot
from matplotlib import font_manager as fmt_manager

import plotly.express as px
import plotly.figure_factory as ff
import plotly.graph_objects as go
from datetime import datetime, timedelta

In [65]:
#reading file
autos = pd.read_csv('autos.csv', encoding="Latin-1")

Inspecting the head and tail of the dataset 

In [67]:
autos.head(5)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [68]:
autos.tail(5)

dateCrawled                                               name  \
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

      brand notRepairedDamage          dateCreated  nrOfPictures  postalCode  \
49995  audi              nein  2016-03-27 00:00:00             0       82131   
49996  opel              nein  2016-03-28 00:00:00             0       44807   
49997  fiat              nein  2016-04-02 00:00:00             0       73430   
49998  audi              nein  2016-03-08 00:00:00             0       35683   
49999  opel              nein  2016-03-13 00:00:00             0       45897   

                  lastSeen  
49995  2016-04-01 13:47:40  
49996  2016-04-02 14:18:02  
49997  2016-04-04 11:47:27  
49998  2016-04-05 16:45:07  
49999  2016-04-06 21:18:48

In [69]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

DATA CLEANING AND TRANSFORMATION

From observation, it is seen that the dataframe contains some null values and the column names are written with the camel case as against the preffered snake. First column name will be converted from camelcase to snakecase and renamed to be more descriptive.

In [70]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We map the corrections to a columm and assign that column back to the dataframe to replace the old column

In [71]:
correction = {'dateCrawled': 'date_crawled',
              'name': 'name',
              'seller': 'seller', 
              'offerType': 'offer_type', 
              'price': 'price', 
              'abtest': 'ab_test',
              'vehicleType': 'vehicle_type', 
              'yearOfRegistration': 'registration_year', 
              'gearbox': 'gearbox', 
              'powerPS': 'power_ps',
              'model': 'model',
              'odometer': "odometer", 
              'monthOfRegistration': 'registration_month', 
              'fuelType': 'fuel_type', 
              'brand': 'brand',
              'notRepairedDamage': 'unrepaired_damage', 
              'dateCreated': 'ad_created', 
              'nrOfPictures': 'nr_of_pictures', 
              'postalCode': 'postal_code',
              'lastSeen': 'last_seen'
}

new_column = autos.columns.map(correction)
autos.columns = new_column
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [72]:
# Using initial descriptive statics for each column in the dataset to explore
autos.describe(include ='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-21 20:37:19  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [73]:
# Taking a quick look at individual columns of interest
autos['gearbox'].value_counts(dropna = False)
autos['price'].value_counts(dropna=False)

$0          1421
$500         781
$1,500       734
$2,500       643
$1,000       639
            ... 
$11,150        1
$173           1
$180,000       1
$459           1
$32,150        1
Name: price, Length: 2357, dtype: int64

The price and odometer columns are numeric data stored as text. We convert them by removing the non numerical characters and convert the values to numeric dtype. In addition we also rename odometer to odometer_km

In [74]:
#replacing the numerical text data in the price column

autos['price'] = autos['price'].str.replace('$','').str.replace(',','').astype(int)
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [75]:
#replacing the numerical text data in the odometer column

autos['odometer'] = autos['odometer'].str.replace('km','').str.replace(',','').astype(int)

autos['odometer'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int64

In [76]:
#rename odometer column

autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True)

autos.describe()

price  registration_year      power_ps    odometer_km  \
count  5.000000e+04       50000.000000  50000.000000   50000.000000   
mean   9.840044e+03        2005.073280    116.355920  125732.700000   
std    4.811044e+05         105.712813    209.216627   40042.211706   
min    0.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.100000e+03        1999.000000     70.000000  125000.000000   
50%    2.950000e+03        2003.000000    105.000000  150000.000000   
75%    7.200000e+03        2008.000000    150.000000  150000.000000   
max    1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        50000.000000         50000.0  50000.000000  
mean             5.723360             0.0  50813.627300  
std              3.711984             0.0  25779.747957  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30451.000000  
50%              6.000000             0.0  49577.000000  
75%              9.000000             0.0  71540.000000  
max             12.000000             0.0  99998.000000

Exploring further, the price column for any patterns or abnormalities

In [77]:
autos['price'].unique().shape

(2357,)

In [78]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [79]:
autos['price'].value_counts().sort_index(ascending=True).head()

0    1421
1     156
2       3
3       1
5       2
Name: price, dtype: int64

Checking the odometer_km column further for any abnormalities or patterns too

In [80]:
autos['odometer_km'].unique().shape

(13,)

In [81]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [82]:
autos['odometer_km'].value_counts().sort_index(ascending=True).head()

5000     967
10000    264
20000    784
30000    789
40000    819
Name: odometer_km, dtype: int64

The max vlaue in the price column is 10^5 times higher in magnitude than 75% of values withing this data set while the odometer column shows a minimum that is 10^2 less than 75% of values in the data set. This is called positive and negative skew respectively. 

REMOVING OUTLIERS

30000 seems to be realistic price range and as such we cut out outliers outside of this range. This figure was chosen on general knowledge based on price of cars. After this we check the new statistical distribution. The same process is repeated for the the odometer column

In [83]:
autos.loc[autos['price'].between(0,30000), 'price'].describe()
#resetting the price column
autos['price'] = autos[autos['price'].between(0,30000)]['price']

In [84]:
# Removing outliers in the odometer_km column and checking its statistical distribution
autos.loc[autos['odometer_km'].between(10000,150000), 'odometer_km'].describe()
#resetting the odometer column after removing outliers
autos['odometer_km'] = autos[autos['odometer_km'].between(10000,150000)]['odometer_km']

EXPLORING DATE

We explore the date columns 'date_crawled', 'ad_created' and 'last_seen' to understand the range of data. We specify the parameters setting normalise to true to get the result in percentage

In [85]:
#exploring date_crawled
autos['date_crawled'].value_counts(normalize=True, dropna=False)

2016-03-21 20:37:19    0.00006
2016-04-02 11:37:04    0.00006
2016-03-30 17:37:35    0.00006
2016-03-23 18:39:34    0.00006
2016-03-16 21:50:53    0.00006
                        ...   
2016-03-07 08:32:20    0.00002
2016-04-01 18:45:44    0.00002
2016-03-09 18:37:55    0.00002
2016-03-20 14:40:53    0.00002
2016-03-27 20:37:58    0.00002
Name: date_crawled, Length: 48213, dtype: float64

In [86]:
date_crawled = autos['date_crawled'].str[:10].value_counts(normalize= True, dropna = False)

In [25]:
date_crawled.sort_index()

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

In [87]:
#exploring ad_created column
ad_created = autos['ad_created'].str[:10].value_counts(normalize= True, dropna = False)
ad_created.sort_index()

2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
               ...   
2016-04-03    0.03892
2016-04-04    0.03688
2016-04-05    0.01184
2016-04-06    0.00326
2016-04-07    0.00128
Name: ad_created, Length: 76, dtype: float64

In [34]:
#exploring last seen
last_seen = autos['last_seen'].str[:10].value_counts(normalize= True, dropna = False)
last_seen.sort_index()

2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64

In [88]:
#exploring the registration with describe method

autos['registration_year'].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The registration year column contains some off values for example the minimum year 1000 and the max year 9999. We can use intuitive knowledge to set a raealistic data range between 1990 and 2016.

In [89]:
autos.loc[-(autos['registration_year'].between(1900, 2016))].describe()
#autos.loc[autos['registration_year'].between(1900,2016), 'registration_year'].describe()

price  registration_year      power_ps    odometer_km  \
count   1963.000000        1972.000000   1972.000000    1916.000000   
mean    3367.427407        2060.353448     98.954361  133987.473904   
std     3955.896136         528.204087    426.692546   32032.014822   
min        0.000000        1000.000000      0.000000   10000.000000   
25%      900.000000        2017.000000      0.000000  125000.000000   
50%     1900.000000        2017.000000     75.000000  150000.000000   
75%     4390.000000        2018.000000    120.000000  150000.000000   
max    29699.000000        9999.000000  16011.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count         1972.000000          1972.0   1972.000000  
mean             4.641988             0.0  47836.475152  
std              3.912620             0.0  25302.249083  
min              0.000000             0.0   1067.000000  
25%              1.000000             0.0  27577.500000  
50%              4.000000             0.0  46006.500000  
75%              8.000000             0.0  66127.750000  
max             12.000000             0.0  99974.000000

We identify the index numbers of the unrealistic years and we can use these index to remove other rows uniformly from the other columns.

In [90]:
index_to_remove = autos.loc[-(autos['registration_year'].between(1900, 2016))].index
index_to_remove

Int64Index([   10,    55,    65,    68,    84,   113,   164,   197,   253,
              348,
            ...
            49662, 49689, 49696, 49731, 49770, 49796, 49841, 49880, 49910,
            49935],
           dtype='int64', length=1972)

We remove index and assign to a variable which becomes our new dataset instead of autos

In [91]:
autos_19_2016 = autos.drop(index_to_remove)
autos_19_2016.describe(include= 'all')

date_crawled         name  seller offer_type         price  \
count                 48028        48028   48028      48028  47243.000000   
unique                46375        37019       2          2           NaN   
top     2016-03-16 21:50:53  Ford_Fiesta  privat    Angebot           NaN   
freq                      3           76   48027      48027           NaN   
mean                    NaN          NaN     NaN        NaN   5094.679635   
std                     NaN          NaN     NaN        NaN   5729.225244   
min                     NaN          NaN     NaN        NaN      0.000000   
25%                     NaN          NaN     NaN        NaN   1100.000000   
50%                     NaN          NaN     NaN        NaN   2900.000000   
75%                     NaN          NaN     NaN        NaN   6999.000000   
max                     NaN          NaN     NaN        NaN  30000.000000   

       ab_test vehicle_type  registration_year  gearbox      power_ps  model  \
count    48028        44903        48028.00000    45604  48028.000000  45560   
unique       2            8                NaN        2           NaN    244   
top       test    limousine                NaN  manuell           NaN   golf   
freq     24756        12859                NaN    35560           NaN   3815   
mean       NaN          NaN         2002.80351      NaN    117.070417    NaN   
std        NaN          NaN            7.31085      NaN    195.151278    NaN   
min        NaN          NaN         1910.00000      NaN      0.000000    NaN   
25%        NaN          NaN         1999.00000      NaN     71.000000    NaN   
50%        NaN          NaN         2003.00000      NaN    107.000000    NaN   
75%        NaN          NaN         2008.00000      NaN    150.000000    NaN   
max        NaN          NaN         2016.00000      NaN  17700.000000    NaN   

          odometer_km  registration_month fuel_type       brand  \
count    47117.000000        48028.000000     44301       48028   
unique            NaN                 NaN         7          40   
top               NaN                 NaN    benzin  volkswagen   
freq              NaN                 NaN     29248       10188   
mean    127874.864699            5.767760       NaN         NaN   
std      36786.714694            3.696802       NaN         NaN   
min      10000.000000            0.000000       NaN         NaN   
25%     125000.000000            3.000000       NaN         NaN   
50%     150000.000000            6.000000       NaN         NaN   
75%     150000.000000            9.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count              39040                48028         48028.0  48028.000000   
unique                 2                   74             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq               34255                 1878             NaN           NaN   
mean                 NaN                  NaN             0.0  50935.867327   
std                  NaN                  NaN             0.0  25792.079828   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  30459.000000   
50%                  NaN                  NaN             0.0  49696.000000   
75%                  NaN                  NaN             0.0  71665.000000   
max                  NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 48028  
unique                38102  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

FURTHER ANALYSIS TO CHECK FOR NULL VALUES

We can check this by logic combination, more like a combined boolean indexing: return values of prices and odometer that are not non null, while checking the other rows using these non null rows as reference.

In [92]:
autos_19_2016.loc[(autos_19_2016['price'].notnull()) & (autos_19_2016['odometer_km'].notnull())].isnull().sum()

date_crawled             0
name                     0
seller                   0
offer_type               0
price                    0
ab_test                  0
vehicle_type          2840
registration_year        0
gearbox               2138
power_ps                 0
model                 2261
odometer_km              0
registration_month       0
fuel_type             3412
brand                    0
unrepaired_damage     8506
ad_created               0
nr_of_pictures           0
postal_code              0
last_seen                0
dtype: int64

REMOVING NULL VALUES

We see that certain columns contain null values for which price and odometer have corresponding values. We will remove these rows with null values for the 'price' and 'odometer_km' columns while also removing the 'unrepaired_damage' column as there is an abundance of null values and the column is of little interest or no significance.

In [93]:
autos_19_2016_cleaner = autos_19_2016.loc[(autos_19_2016['price'].notnull()) & (autos_19_2016['odometer_km'].notnull())]
autos_19_2016_cleaner.drop(['unrepaired_damage'], axis = 1, inplace = True)
autos_19_2016_cleaner.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46378 entries, 0 to 49999
Data columns (total 19 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   date_crawled        46378 non-null  object 
 1   name                46378 non-null  object 
 2   seller              46378 non-null  object 
 3   offer_type          46378 non-null  object 
 4   price               46378 non-null  float64
 5   ab_test             46378 non-null  object 
 6   vehicle_type        43538 non-null  object 
 7   registration_year   46378 non-null  int64  
 8   gearbox             44240 non-null  object 
 9   power_ps            46378 non-null  int64  
 10  model               44117 non-null  object 
 11  odometer_km         46378 non-null  float64
 12  registration_month  46378 non-null  int64  
 13  fuel_type           42966 non-null  object 
 14  brand               46378 non-null  object 
 15  ad_created          46378 non-null  object 
 16  nr_o

ANALYSIS 

We want to aggregate on the brand column by identifying unique values and counting the unique values.

In [94]:
#creating an empty dictionary
#autos_19_2016_cleaner['brand'].value_counts().sort_index
#brand_dict = {}
autos_19_2016_cleaner['brand'].nunique()
    

40

Estimating top 20 selling brands

In [95]:
auto_brands_20 = autos_19_2016_cleaner['brand'].value_counts().head(20).index
auto_brands_20

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'smart',
       'citroen', 'toyota', 'hyundai', 'volvo', 'sonstige_autos', 'mini'],
      dtype='object')

Estimating the mean selling price for the top 20 selling car brands that will be saved to a dictionary

In [96]:
#create dictionary
brand_mean_prices = {}
for brand in auto_brands_20:
    mean_price = round(autos_19_2016_cleaner.loc[autos_19_2016_cleaner['brand'] == brand, 'price'].mean())
    # Creating a new key_value pair with this mean price
    brand_mean_prices[brand] = mean_price

In [39]:
brand_mean_prices

{'volkswagen': 5038.0,
 'opel': 2859.0,
 'bmw': 7247.0,
 'mercedes_benz': 7357.0,
 'audi': 7773.0,
 'ford': 3407.0,
 'renault': 2319.0,
 'peugeot': 3044.0,
 'fiat': 2674.0,
 'seat': 4265.0,
 'skoda': 6271.0,
 'mazda': 3815.0,
 'nissan': 4607.0,
 'smart': 3561.0,
 'citroen': 3621.0,
 'toyota': 4989.0,
 'hyundai': 5282.0,
 'volvo': 4781.0,
 'sonstige_autos': 7417.0,
 'mini': 10334.0}

In [97]:
# sorting the dictionary key values 
brand_mean_prices = dict(sorted(brand_mean_prices.items(), key = lambda data: data[1], reverse = True ))
brand_mean_prices

{'mini': 10334.0,
 'audi': 7773.0,
 'sonstige_autos': 7417.0,
 'mercedes_benz': 7357.0,
 'bmw': 7247.0,
 'skoda': 6271.0,
 'hyundai': 5282.0,
 'volkswagen': 5038.0,
 'toyota': 4989.0,
 'volvo': 4781.0,
 'nissan': 4607.0,
 'seat': 4265.0,
 'mazda': 3815.0,
 'citroen': 3621.0,
 'smart': 3561.0,
 'ford': 3407.0,
 'peugeot': 3044.0,
 'opel': 2859.0,
 'fiat': 2674.0,
 'renault': 2319.0}

From sorting the dictionary values, in descending order, the highest selling prices among top 20 was mini, audi and mercedes while renault, fiat and opel were least expenses among the top 20. The price gap between the highest and lowest selling is almost 2.5 times. Mazda and Citreon are mid range priced cars.

Aggregating top 6 brands using the Series constructor method

In [98]:
brand_mean_prices = {}
brand_mean_mileage = {}
for brand in auto_brands_20:
    mean_price = round(autos_19_2016_cleaner.loc[autos_19_2016_cleaner['brand'] == brand, 'price'].mean())
    mean_mileage = round(autos_19_2016_cleaner.loc[autos_19_2016_cleaner['brand'] == brand, 'odometer_km'].mean())
    # Creating a new key_value pair with this mean price and mean mileage
    brand_mean_prices[brand] = mean_price
    brand_mean_mileage[brand] = mean_mileage

In [99]:
brand_mean_prices

{'volkswagen': 5038.0,
 'opel': 2859.0,
 'bmw': 7247.0,
 'mercedes_benz': 7357.0,
 'audi': 7773.0,
 'ford': 3407.0,
 'renault': 2319.0,
 'peugeot': 3044.0,
 'fiat': 2674.0,
 'seat': 4265.0,
 'skoda': 6271.0,
 'mazda': 3815.0,
 'nissan': 4607.0,
 'smart': 3561.0,
 'citroen': 3621.0,
 'toyota': 4989.0,
 'hyundai': 5282.0,
 'volvo': 4781.0,
 'sonstige_autos': 7417.0,
 'mini': 10334.0}

In [100]:
brand_mean_mileage

{'volkswagen': 131497.0,
 'opel': 131987.0,
 'bmw': 136564.0,
 'mercedes_benz': 134396.0,
 'audi': 133746.0,
 'ford': 126649.0,
 'renault': 130547.0,
 'peugeot': 128355.0,
 'fiat': 119694.0,
 'seat': 123463.0,
 'skoda': 113040.0,
 'mazda': 127170.0,
 'nissan': 121051.0,
 'smart': 100309.0,
 'citroen': 121002.0,
 'toyota': 117314.0,
 'hyundai': 107596.0,
 'volvo': 139568.0,
 'sonstige_autos': 103481.0,
 'mini': 90829.0}

In [101]:
# sorting the values 
brand_mean_prices = dict(sorted(brand_mean_prices.items(), key = lambda data: data[1], reverse = True ))
brand_mean_mileage = dict(sorted(brand_mean_mileage.items(), key = lambda data: data[1], reverse = True ))

In [102]:
brand_mean_prices

{'mini': 10334.0,
 'audi': 7773.0,
 'sonstige_autos': 7417.0,
 'mercedes_benz': 7357.0,
 'bmw': 7247.0,
 'skoda': 6271.0,
 'hyundai': 5282.0,
 'volkswagen': 5038.0,
 'toyota': 4989.0,
 'volvo': 4781.0,
 'nissan': 4607.0,
 'seat': 4265.0,
 'mazda': 3815.0,
 'citroen': 3621.0,
 'smart': 3561.0,
 'ford': 3407.0,
 'peugeot': 3044.0,
 'opel': 2859.0,
 'fiat': 2674.0,
 'renault': 2319.0}

In [103]:
brand_mean_mileage

{'volvo': 139568.0,
 'bmw': 136564.0,
 'mercedes_benz': 134396.0,
 'audi': 133746.0,
 'opel': 131987.0,
 'volkswagen': 131497.0,
 'renault': 130547.0,
 'peugeot': 128355.0,
 'mazda': 127170.0,
 'ford': 126649.0,
 'seat': 123463.0,
 'nissan': 121051.0,
 'citroen': 121002.0,
 'fiat': 119694.0,
 'toyota': 117314.0,
 'skoda': 113040.0,
 'hyundai': 107596.0,
 'sonstige_autos': 103481.0,
 'smart': 100309.0,
 'mini': 90829.0}

The benz audi and opel seem to have the most mileage of the cars while the mini, smart and sonstige autos have the lowest mileage of the top 20 brands. We use the Series and Df constrcutor to aggragate these two column variables.
We convert both dictioanries to Series objects (mean price & mileage dictionary)

In [104]:
mp_series = pd.Series(brand_mean_prices)
mileage_series = pd.Series(brand_mean_mileage)

print(mp_series)
print(mileage_series)

mini              10334.0
audi               7773.0
sonstige_autos     7417.0
mercedes_benz      7357.0
bmw                7247.0
skoda              6271.0
hyundai            5282.0
volkswagen         5038.0
toyota             4989.0
volvo              4781.0
nissan             4607.0
seat               4265.0
mazda              3815.0
citroen            3621.0
smart              3561.0
ford               3407.0
peugeot            3044.0
opel               2859.0
fiat               2674.0
renault            2319.0
dtype: float64
volvo             139568.0
bmw               136564.0
mercedes_benz     134396.0
audi              133746.0
opel              131987.0
volkswagen        131497.0
renault           130547.0
peugeot           128355.0
mazda             127170.0
ford              126649.0
seat              123463.0
nissan            121051.0
citroen           121002.0
fiat              119694.0
toyota            117314.0
skoda             113040.0
hyundai           107596.0
sonsti

In [105]:
# creating a dataframe from the first series (mean_price)
brands_top20 = pd.DataFrame(mp_series, columns=['mean_price'])
brands_top20

mean_price
mini               10334.0
audi                7773.0
sonstige_autos      7417.0
mercedes_benz       7357.0
bmw                 7247.0
skoda               6271.0
hyundai             5282.0
volkswagen          5038.0
toyota              4989.0
volvo               4781.0
nissan              4607.0
seat                4265.0
mazda               3815.0
citroen             3621.0
smart               3561.0
ford                3407.0
peugeot             3044.0
opel                2859.0
fiat                2674.0
renault             2319.0

In [106]:
# assign the other series(mileage) as a new column in the dataframe above

#brands_top20.assign(column=mileage_series)

brands_top20['mean_mileage_km'] =  mileage_series 

In [72]:
# sorting the mean mileage values, since the one for price has been sorted akready
brands_top20.sort_values('mean_mileage_km', ascending = False)

mean_price  mean_mileage_km
volvo               4781.0         139568.0
bmw                 7247.0         136564.0
mercedes_benz       7357.0         134396.0
audi                7773.0         133746.0
opel                2859.0         131987.0
volkswagen          5038.0         131497.0
renault             2319.0         130547.0
peugeot             3044.0         128355.0
mazda               3815.0         127170.0
ford                3407.0         126649.0
seat                4265.0         123463.0
nissan              4607.0         121051.0
citroen             3621.0         121002.0
fiat                2674.0         119694.0
toyota              4989.0         117314.0
skoda               6271.0         113040.0
hyundai             5282.0         107596.0
sonstige_autos      7417.0         103481.0
smart               3561.0         100309.0
mini               10334.0          90829.0

It can be seen that the cars with lower mileage sell higher than cars with higher mileage. looking at the exploration below, it is seen that the Mini brand has the max selling price as well as the lowest mileage. 

In [107]:
# Exploring the data simoltaneously to check if any relationship between 'price' and 'mileage' if any

print(brands_top20['mean_price'].idxmax())
print(brands_top20['mean_price'].idxmin())


mini
renault


In [108]:
# Exploring the data simoltaneously to check if any relationship between 'price' and 'mileage'
print(brands_top20['mean_mileage_km'].idxmax())
print(brands_top20['mean_mileage_km'].idxmin())

volvo
mini


CONCLUSION

To conclude the project, we will identify a few categorical data like the offer type column and seller and translate the contents from German to English. Will also write the data to a new csv to use for visualization in Tableau or Power Bi.

In [109]:
autos_19_2016_cleaner.head(4)
autos_19_2016_cleaner.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand', 'ad_created',
       'nr_of_pictures', 'postal_code', 'last_seen'],
      dtype='object')

In [110]:
autos_19_2016_cleaner['offer_type'].unique()

array(['Angebot', 'Gesuch'], dtype=object)

In [111]:
translate_1 = {'Angebot': 'Offer',
              'Gesuch': 'Search'
}

translate_2 = {'privat': 'Personal',
              'gewerblich': 'Business'
}

translate_3 = {'Angebot': 'Offer',
              'Gesuch': 'Search'
}

#autos_19_2016_cleaner['offer_type'] = autos_19_2016_cleaner['offer_type'].map(translate)

#autos_19_2016_cleaner['seller'] = autos_19_2016_cleaner['seller'].map(translate)

autos_19_2016_cleaner['offer_type']

0        Offer
1        Offer
2        Offer
3        Offer
4        Offer
         ...  
49994    Offer
49995    Offer
49996    Offer
49998    Offer
49999    Offer
Name: offer_type, Length: 46378, dtype: object

In [120]:
autos_19_2016_cleaner.to_csv('ebay_cars.csv', index=False)